contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1625 @@
+        KLIMA = _KLIMA;
+        sKLIMA = _sKLIMA;
+        wsKLIMA = _wsKLIMA;
+        USDC = _USDC;
+        staking = _staking;
+        stakingHelper = _stakingHelper;
+        DAO = _DAO;
+        KlimaRetirementStorage = _KlimaRetirementStorage;
+    }
+
+    address public KLIMA;
+    address public sKLIMA;
+    address public wsKLIMA;
+    address public immutable USDC;
+    address public staking;
+    address public stakingHelper;
+    address public immutable DAO;
+    address public KlimaRetirementStorage;
+
+    // Event setup
+    event ToucanRetired(
+        address indexed retiringAddress,
+        address indexed beneficiaryAddress,
+        string beneficiaryString,
+        string retirementMessage,
+        address indexed carbonPool,
+        address carbonToken,
+        uint256 retiredAmount
+    );
+    event PoolAdded(
+        address indexed carbonPool,
+        address indexed poolRouter,
+        address indexed bondDepository
+    );
+    event PoolRouterChanged(
+        address indexed carbonPool,
+        address indexed oldRouter,
+        address indexed newRouter
+    );
+    event PoolBondChanged(
+        address indexed carbonPool,
+        address indexed oldBondDepository,
+        address indexed newBondDepository
+    );
+    event FeeUpdated(uint256 oldFee, uint256 newFee);
+
+    address[] public toucanPools;
+    mapping(address => bool) public isToucanPool;
+    mapping(address => address) public poolRouter;
+    mapping(address => address) public bondDepository;
+
+    uint256 feeAmount;
+
+    /**
+        @notice Get KLIMA type provided to regular KLIMA, swap, then retire.
+        @param _klimaType Address of the original KLIMA token provided (Wrapped/Staked/Unstaked).
+        @param _amount Total pool tokens being retired. Expected uint with 18 decimals.
+        @param _beneficiaryAddress Address of the beneficiary if different than sender. Value is set to msg.sender if null is sent.
+        @param _beneficiaryString String that can be used to describe the beneficiary
+        @param _retirementMessage String for specific retirement message if needed.
+        @param _listTCO2 List of TCO2 token addresses to redeem pool tokens.
+        @param _toucanPool Address of pool token being used to retire.
+     */
+    function retireWithKLIMA(
+        address _klimaType,
+        uint256 _amount,
+        address _beneficiaryAddress,
+        string calldata _beneficiaryString,
+        string calldata _retirementMessage,
+        address[] calldata _listTCO2,
+        address _toucanPool
+    ) public {
+        require(isToucanPool[_toucanPool], "Not a Toucan Protocol Carbon Pool");
+
+        uint256 totalPoolNeeded = (_amount * (1000 + feeAmount)) / 1000;
+
+        (uint256 amountIn, address[] memory path) = getNeededBuyAmount(
+            KLIMA,
+            _toucanPool,
+            totalPoolNeeded
+        );
+
+        uint256 unwrappedKLIMA = _stakedToUnstaked(_klimaType, amountIn);
```

Comment:
This does not need to be executed when type is KLIMA so a cleaner approach may be to do the following:
```js
if (_klimaType == KLIMA) {
    IERC20(KLIMA).safeTransferFrom(msg.sender, address(this), amountIn);
} else {
    amountIn = _stakedToUnstaked(_klimaType, amountIn);
}
```
Then you can also drop the extra `_unwrappedKLIMA` variable from `_returnKLIMADust` as it's the same as `amountIn`.